File: ml-benchmark/synduce/treepaths/height.f
Inductive BTree = empty Unit | node {Int, BTree, BTree};
Inductive Zipper = top Unit | left {Int, BTree, Zipper} | right {Int, BTree, Zipper};

max = \a: Int. \b: Int. if (< a b) then b else a;

height = fix (
  \f: BTree -> Int. \t: BTree.
  match t with
    empty _ -> 0
  | node {a, l, r} -> + 1 (max (f l) (f r))
  end
);

repr = fix (
  \f: Zipper -> Compress BTree. \z: Zipper.
  match z with
    top _ -> empty unit
  | left {w, tree, zz} ->
    /* (height tree) is given in Synduce's template*/
    let tw = height tree in
      node {w, tree, f zz}
  | right {w, tree, zz} ->
    let tw = height tree in
      node {w, f zz, tree}
  end
);

main = \z: Zipper. height (repr z);

File: ml-benchmark/synduce/treepaths/height.ml
type btree = Empty of unit | Node of int * btree * btree

type zipper = Top of unit | Left of int * btree * zipper | Right of int * btree * zipper

let max a b = if a < b then b else a

let rec height t =
  match t with
  | Empty _ -> 0
  | Node (a, l, r) -> 1 + max (height l) (height r)

val repr: zipper -> btree compress
let rec repr z =
  match z with
  | Top _ -> Empty ()
  | Left (w, tree, zz) ->
      let tw = height tree in
      Node (w, tree, repr zz)
  | Right (w, tree, zz) ->
      let tw = height tree in
      Node (w, repr zz, tree)

let program z = height (repr z)